### PR TITLE
[MO] cli_parser fix when input contains substring with matching scale/mean values

### PR DESCRIPTION
**Pull request by pavel-esir**
Description:  When input contained substring that matches scale/mean values e.g. 

> --scale_values data.70[127.5],data.**1**[**1**],

cli_parses incorrectly handles that case. This PR fixes this issue.

ticket #42774

Code:
* [x]  Comments
* [x]  Code style (PEP8)
* [x]  Transformation generates reshape-able IR: N/A -- no new transformations were added
* [x]  Transformation preserves original framework node names: N/A -- because no transformation were added

Validation:
* [x]  Unit tests:
* [x]  Framework operation tests: N/A -- no new operations were added
* [x]  Transformation tests: N/A -- no new transformations were added
* [x]  e2e model test with an update of ./tests/e2e_oss/utils/reshape_utils.py: N/A
* [x]  Model Optimizer IR Reader check

Documentation:
* [x]  Supported frameworks operations list: N/A
* [x]  Supported **public** models list: N/A
* [x]  New operations specification: N/A
* [x]  Guide on how to convert the **public** model: N/A
* [x]  User guide update: N/A

### CODE REVIEW

**Comment by lazarevevgeny** [comment]
Based on the changed code size and the issue priority I change the milestone to 2021.3.

File: model-optimizer/mo/utils/cli_parser_test.py [hunk 1]
```diff
@@ -355,6 +355,15 @@ def test_mean_do_not_match_input(self):
         mean_values = parse_tuple_pairs("input_not_present(255),input2(255)")
         self.assertRaises(Error, get_mean_scale_dictionary, mean_values, scale_values, "input1,input2")
 
+    def test_values_match_input_name(self):
+        # before fix cli_parser was confused when input had a substring that matches scale/mean values
+        values = parse_tuple_pairs("input255(255),input255.0(255.0)")
+        exp_res = {'input255': 255.0, 'input255.0': 255.0}
+        self.assertEqual(exp_res, values)
```
Comment (rkazants) [inline model-optimizer/mo/utils/cli_parser_test.py:362]: I would also want you to cover cases:
1. square brackets - input255.0[1.0,2.0,3.0];
2. several dots - input255.0.3[2.0,3.0];
3. several inputs - input.1[2.0,3.0],input.2[4.0,5.0],;

Comment (pavel-esir) [reply model-optimizer/mo/utils/cli_parser_test.py]: @rkazants, thanks for your feedback!
1. Luckily, square brackets are covered in test_multi_mean_scale_arrays_no_input ("[1.1,22.22,333.333],[2.1,33.22,444.333]"); test_multi_mean_scale_input ("data[1.1,22.22,333.333],info[2.1,33.22,444.333]")
2. algorithm is not bound to dots outside of brackets so inputs with such names are safely processed. Added one test case to reduce change that someone in future will break it
3. added input with several dots in name to the same test so now there is 3 inputs but this does not change anything in principle -- 2 inputs are processed same as 4, 5 and etc.

Also added some code for deeper test result comparison. Please review.

File: model-optimizer/mo/utils/cli_parser_test.py [hunk 2]
```diff
@@ -355,6 +355,20 @@ def test_mean_do_not_match_input(self):
         mean_values = parse_tuple_pairs("input_not_present(255),input2(255)")
         self.assertRaises(Error, get_mean_scale_dictionary, mean_values, scale_values, "input1,input2")
 
+    def test_values_match_input_name(self):
+        # before fix cli_parser was confused when input had a substring that matches scale/mean values
+        res_values = parse_tuple_pairs("input255(255),input255.0(255.0),multi-dotted.input.3.(255,128,64)")
+        exp_res = {'input255': np.array([255.0, 254.]),
```
Comment (rkazants) [inline model-optimizer/mo/utils/cli_parser_test.py:361]: Why do we have 254.0 and how do the tests pass?

File: model-optimizer/mo/utils/cli_parser_test.py [hunk 3]
```diff
@@ -355,6 +355,20 @@ def test_mean_do_not_match_input(self):
...
+                   'input255.0': np.array([255.0]),
+                   'multi-dotted.input.3.': np.array([255., 128., 128.])}
+        self.assertEqual(len(exp_res), len(res_values))
+        for i, j in zip(exp_res, res_values):
+            self.assertEqual(i, j)
+            np.array_equal(exp_res[i], res_values[j])
```
Comment (rkazants) [inline model-optimizer/mo/utils/cli_parser_test.py:367]: you need to insert assert for `array_equal`.

**Review by rkazants — CHANGES_REQUESTED** [review]
see my comments

File: model-optimizer/mo/utils/cli_parser_test.py (hunk 2 above)
Comment (pavel-esir) [reply model-optimizer/mo/utils/cli_parser_test.py:361]: Thanks for noting this. Agree.

File: model-optimizer/mo/utils/cli_parser_test.py (hunk 3 above)
Comment (pavel-esir) [reply model-optimizer/mo/utils/cli_parser_test.py:367]: Agree.

**Comment by pavel-esir** [comment]
Corrected unit-test values, added 'assert_' to numpy array comparison. @rkazants, please review.

**Review by rkazants — APPROVED** [review]

File: model-optimizer/mo/utils/cli_parser.py [hunk 4]
```diff
@@ -917,43 +917,28 @@ def parse_tuple_pairs(argv_values: str):
...
-                        "Mean/scale values should either contain names of input layers: data(1,2,3),info(2,3,4)" +
-                        " or just plain set of them without naming any inputs: (1,2,3),(2,3,4)." +
-                        refer_to_faq_msg(101), argv_values)
-            for match in tuples_matches:
-                res.append(np.fromstring(match, dtype=float, sep=','))
-            break
+    matches = [m for m in re.finditer(r'[(\[]([0-9., -]+)[)\]]', argv_values, re.IGNORECASE)]
 
-        res[input_name] = np.fromstring(tuple_value, dtype=float, sep=',')
+    error_msg = "Mean/scale values should be in format: data(1,2,3),info(2,3,4)" \
```
Comment (lazarevevgeny) [inline model-optimizer/mo/utils/cli_parser.py:922]: should we update the message that `[]` are also supported?

File: model-optimizer/mo/utils/cli_parser.py [hunk 5]
```diff
@@ -917,43 +917,28 @@ def parse_tuple_pairs(argv_values: str):
...
+    for m in matches:
+        input_name = argv_values[name_start_idx:m.start(0)]
+        name_start_idx = m.end(0) + 1
+        tuple_value = m.groups()[0]
 
-        if not data_str:
-            break
+        if not input_name:
+            # check that other values are specified without names
+            if re.search(r'([a-zA-Z]+)', argv_values[m.start(0):]) is None:
```
Comment (lazarevevgeny) [inline model-optimizer/mo/utils/cli_parser.py:936]: what if the name consists of digits only? It is quite common for ONNX models. Please, add a unit test to check that case.

File: model-optimizer/mo/utils/cli_parser_test.py [hunk 6]
```diff
@@ -355,6 +356,20 @@ def test_mean_do_not_match_input(self):
         mean_values = parse_tuple_pairs("input_not_present(255),input2(255)")
         self.assertRaises(Error, get_mean_scale_dictionary, mean_values, scale_values, "input1,input2")
 
+    def test_values_match_input_name(self):
+        # before fix cli_parser was confused when input had a substring that matches scale/mean values
```
Comment (lazarevevgeny) [inline model-optimizer/mo/utils/cli_parser_test.py:360]: Please, refactor the comment. It is not clear what "fix" you are referring to.

**Review by lazarevevgeny — CHANGES_REQUESTED** [review]

File: model-optimizer/mo/utils/cli_parser.py (hunk 5 above)
Comment (pavel-esir) [reply model-optimizer/mo/utils/cli_parser.py]: Thanks for very reasonable question! Old solution was not processing correctly names consisting of digits. As well as situations when node name firstly was specified and then not (e.g. `input(255),[127]`). Did some improvements. Now we catch them. Please review.

File: model-optimizer/mo/utils/cli_parser.py (hunk 4 above)
Comment (pavel-esir) [reply model-optimizer/mo/utils/cli_parser.py]: Updated the message

File: model-optimizer/mo/utils/cli_parser_test.py (hunk 6 above)
Comment (pavel-esir) [reply model-optimizer/mo/utils/cli_parser_test.py]: Changed this comment to a more meaningful

File: model-optimizer/mo/utils/cli_parser.py [hunk 7]
```diff
@@ -917,44 +917,33 @@ def parse_tuple_pairs(argv_values: str):
...
+        input_name = argv_values[name_start_idx:match.start(0)]
+        name_start_idx = match.end(0) + 1
+        tuple_value = np.fromstring(match.groups()[0], dtype=float, sep=',')
+
+        if idx != 0 and name_was_present ^ bool(input_name):
+            # if node name firstly was specified and then subsequently not or vice versa
+            # e.g. (255),input[127] or input(255),[127]
+            raise Error(error_msg, argv_values)
+
+        name_was_present = True if input_name else False
```
Comment (lazarevevgeny) [inline model-optimizer/mo/utils/cli_parser.py:940]: `name_was_present = input_name != ""` ?

Comment (pavel-esir) [reply model-optimizer/mo/utils/cli_parser.py:940]: Yes, this line does `name_was_present = input_name != ""`

Comment (lazarevevgeny) [reply model-optimizer/mo/utils/cli_parser.py:940]: I recommend to change the code like I proposed. For me it is more intuitive rather then `name_was_present = True if input_name else False`. However, I don't insist.

**Comment by lazarevevgeny** [comment]
Some unit tests are failing now.

File: model-optimizer/mo/utils/cli_parser.py [hunk 8]
```diff
@@ -917,44 +917,33 @@ def parse_tuple_pairs(argv_values: str):
...
+        raise Error(error_msg, argv_values)
+
+    name_start_idx = 0
+    name_was_present = False
+    for idx, match in enumerate(matches):
+        input_name = argv_values[name_start_idx:match.start(0)]
+        name_start_idx = match.end(0) + 1
+        tuple_value = np.fromstring(match.groups()[0], dtype=float, sep=',')
+
+        if idx != 0 and name_was_present ^ bool(input_name):
```
Comment (lazarevevgeny) [inline model-optimizer/mo/utils/cli_parser.py:935]: please, add parentheses around `name_was_present ^ bool(input_name)` to be precise in the condition logic.

File: model-optimizer/mo/utils/cli_parser.py (hunk 7 above)
Comment (pavel-esir) [reply model-optimizer/mo/utils/cli_parser.py:940]: Agree. Firstly I also had written `input_name != ""` but then googled and most of others use `if input_name` so I changed to more conventional. But `input_name != ""` is more explicit. Changed to it.

File: model-optimizer/mo/utils/cli_parser.py (hunk 8 above)
Comment (pavel-esir) [reply model-optimizer/mo/utils/cli_parser.py]: Done

**Review by lazarevevgeny — APPROVED** [review]

**Pull request merged**